In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler

%matplotlib inline

df = pd.read_csv('train.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [2]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [3]:
print("중복된 항목 수 :", len(df[df.duplicated()])) 

중복된 항목 수 : 0


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [5]:
# categorical and numeric columns
for column_name in list(df.columns):
    print(column_name, df[column_name].dtype, df[column_name].unique())

PassengerId int64 [  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180
 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198
 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216
 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234
 235 236 237 238 239 240 241 242 

In [6]:
'''
  To do 
  위의 결과를 기반으로, categorical data들을 모두 object type으로 변경
'''

In [7]:
'''
	To do
	아래 print문에 맞게 코드를 작성 
'''

print("all columns : ", len(df.columns))
print("number of categorical columns : ", len(list_categorical_columns))
print("number of numerical columns : ", len(list_numeric_columns))

all columns :  12
number of categorical columns :  7
number of numerical columns :  5


In [8]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

#### Summary of Data Set
##### - Number of Variables : 12
    categorical variables : 7
    numeric variables : 5
##### - Number of observations : 891 (duplicated 데이터 없음)
##### - Missing Cells : 3 columns

### 0. Remove unnecessary columns

In [9]:
'''
	To do
	Unique value 숫자 확인하기. 
'''


Sex              2
Pclass           3
Embarked         3
Cabin          147
Ticket         681
PassengerId    891
Name           891
dtype: int64

In [10]:
list_removed_column = ["Name", "PassengerId", "Ticket", "Cabin"]
df = df.drop(list_removed_column, axis=1)
for ele in list_removed_column:
    list_categorical_columns.remove(ele)

In [11]:
df[list_numeric_columns].nunique().sort_values()

Survived      2
SibSp         7
Parch         7
Age          88
Fare        248
dtype: int64

In [12]:
label = df["Survived"]
df = df.drop("Survived", axis=1)
list_numeric_columns.remove("Survived")

### 1. FillNA

In [13]:
df.isna().sum()

Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

In [14]:
'''
	To do
	Age -> 평균값으로 결측치 채우기
    Embarked -> 'N" 으로 결측치 채우기 
'''


In [15]:
df.isna().sum()

Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

### 2. Scaling

In [16]:
scaler = MinMaxScaler()
df.loc[:, list_numeric_columns] = scaler.fit_transform(df[list_numeric_columns])
df.head(3)

Pclass     Sex       Age  SibSp  Parch      Fare Embarked
0      3    male  0.271174  0.125    0.0  0.014151        S
1      1  female  0.472229  0.125    0.0  0.139136        C
2      3  female  0.321438  0.000    0.0  0.015469        S

In [17]:
'''
	To do
	one hot encoding 진행하기 
'''


/var/folders/ns/7ygp82v92rl4_clvg202fw4m0000gn/T/ipykernel_1874/66388900.py:5: FutureWarning: In a future version, the Index constructor will not infer numeric dtypes when passed object-dtype sequences (matching Series behavior)
  df = pd.get_dummies(df)


In [18]:
df

Age  SibSp     Parch      Fare  Pclass_1  Pclass_2  Pclass_3  \
0    0.271174  0.125  0.000000  0.014151         0         0         1   
1    0.472229  0.125  0.000000  0.139136         1         0         0   
2    0.321438  0.000  0.000000  0.015469         0         0         1   
3    0.434531  0.125  0.000000  0.103644         1         0         0   
4    0.434531  0.000  0.000000  0.015713         0         0         1   
..        ...    ...       ...       ...       ...       ...       ...   
886  0.334004  0.000  0.000000  0.025374         0         1         0   
887  0.233476  0.000  0.000000  0.058556         1         0         0   
888  0.367921  0.125  0.333333  0.045771         0         0         1   
889  0.321438  0.000  0.000000  0.058556         1         0         0   
890  0.396833  0.000  0.000000  0.015127         0         0         1   

     Sex_female  Sex_male  Embarked_C  Embarked_N  Embarked_Q  Embarked_S  
0             0         1           0           0           0           1  
1             1         0           1           0           0           0  
2             1         0           0           0           0           1  
3             1         0           0           0           0           1  
4             0         1           0           0           0           1  
..          ...       ...         ...         ...         ...         ...  
886           0         1           0           0           0           1  
887           1         0           0           0           0           1  
888           1         0           0           0           0           1  
889           0         1           1           0           0           0  
890           0         1           0           0           1           0  

[891 rows x 13 columns]

In [19]:
df["Survived"] = label

### 3. Save Dataset

In [20]:
df.to_csv("dataset_preprocessing.csv")